### ISSUE TITLE
setSaveAndRecall(false) does not prevent metaparameter recall

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#266]
Not sure what to do about this. I want my smoothingTime knob to be automatable (so it needs to be attached to a metaparameter), but I don't want the metaparameter value in the preset to be used when changing presets. Basically, only the host or user should be able to change smoothing value.

Edit: Maybe it's as simple as stopping metaparameter from storing the value in the xml. But the problem is that it will load a default value (0.5) if no value exists.

Edit: Btw, we don't need to stop the metaparameter value from loading. If it's more simple to just make sure it doesn't store it in the xml file, that's fine too.

**Comment by elanhickler** [comment]
temporary solution
![image](https://user-images.githubusercontent.com/9116575/50612092-11eae200-0e96-11e9-90d8-7415e5250929.png)

**Comment by RobinSchmidt** [comment]
>  just make sure it doesn't store it in the xml file, that's fine too.

but what if other parameters are also attached to the same meta-parameter? soo - your current solution retrieves the smoothing value, then does the regular state recall and then restores the smoothing value again. i guess that will then also set the meta-value back to its old value, keeping it in sync with the attached smoothing parameter, right?

**Comment by elanhickler** [comment]
Found a bug with your metaparameter preset system.

If the parameter value is defined in the xml but the metaparameter value is not defined, the parameter value gets a normalized 0.5 value since that is the default metaparameter value.

Instead.

It needs to get the default normalized value for the parameter in question, NOT the default 0.5 normalized value of the metaparameter.

**Comment by elanhickler** [comment]
> > just make sure it doesn't store it in the xml file, that's fine too.
> 
> but what if other parameters are also attached to the same meta-parameter? soo - your current solution retrieves the smoothing value, then does the regular state recall and then restores the smoothing value again. i guess that will then also set the meta-value back to its old value, keeping it in sync with the attached smoothing parameter, right?

uhhh... users shouldn't be attaching parameters to smoothing.

**Comment by RobinSchmidt** [comment]
> If the parameter value is defined in the xml but the metaparameter value is not defined, the parameter value gets a normalized 0.5 value since that is the default metaparameter value.

hmm...that sounds more like the bug is that the meta is not defined. i mean, if a meta has an attached parameter, that meta *should be saved* into the xml (unless it's 0.5). the situation that this meta is undefined in the xml should not occur

> Instead, it needs to get the default normalized value for the parameter in question, NOT the default 0.5 normalized value of the metaparameter.

why should the parameter be set to its default when it has a value in the xml? i might be mistaken - it's a while ago

**Comment by RobinSchmidt** [comment]
hmm - ok - but when the preset files are manually edited, i think it should not really be considered a bug in the system.  i have this code (irrelevant bits deleted) where i already made a comment about another situation that could occur on the developer side:
```cpp
void AudioModule::setStateFromXml(const XmlElement& xml, const juce::String& stateName,
  bool markAsClean)
{
  recallParametersFromXml(convertedState);
  if(metaParamManager != nullptr)
  {
    recallMetaMappingFromXml(convertedState);
    recallMetaValuesFromXml(convertedState);
    // maybe this should be done before retrieving the parameter values - because when you change
    // a parameter range in the code and then load an older preset, a parameter that is connected
    // to a meta parameter will get the wrong value.
  }
}
```
i could try to reverse the order of parameter recall and meta recall. this could solve your problem and at the same time the one i anticipated in my comment. ....but: that may be dangerous - i have no idea, if that could have side effects which break something else. although i also don't really see, why it should. we could try it - but then thorough testing would be in order

**Comment by RobinSchmidt** [comment]
one thing i can anticipate is this: when you connect two parameters two the same meta, save the patch, then change the parameter range in the code, then load the patch again, one of the two parameters would get the wrong value (i think). if you only have the situation where one parameter is connected to one meta, it should have the desired behavior (that the range change in the code doesn't affect the value of the recalled parameter - unless the new range is narrower than the old and the stored value is outside the new range)

**Comment by elanhickler** [comment]
no longer relevant
